src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs

```diff
@@ -0,0 +1,403 @@
+
+            var resourceData = pdEvent.Url.OriginalString.EndsWith(".nupkg", StringComparison.OrdinalIgnoreCase)
+                ? data.Nupkg
+                : data.Metadata;
+
+            lock (resourceData.Lock)
+            {
+                ApplyTiming(resourceData.EventTiming, pdEvent.EventDuration);
+
+                if (pdEvent.HeaderDuration.HasValue)
+                {
+                    if (resourceData.HeaderTiming == null)
+                    {
+                        resourceData.HeaderTiming = new ResourceTimingData();
+                    }
+
+                    ApplyTiming(resourceData.HeaderTiming, pdEvent.HeaderDuration.Value);
+                }
+
+                if (pdEvent.IsSuccess)
+                {
+                    resourceData.Successful++;
+                }
+
+                if (pdEvent.IsRetry)
+                {
+                    resourceData.Retries++;
+                }
+
+                if (pdEvent.IsCancelled)
+                {
+                    resourceData.Cancelled++;
+                }
+
+                if (pdEvent.IsLastAttempt && !pdEvent.IsSuccess && !pdEvent.IsCancelled)
+                {
+                    resourceData.Failed++;
+                }
+
+                if (pdEvent.Bytes > 0)
+                {
+                    resourceData.TotalBytes += pdEvent.Bytes;
+                    if (pdEvent.Bytes > resourceData.MaxBytes)
+                    {
+                        resourceData.MaxBytes = pdEvent.Bytes;
+                    }
+                }
+
+                if (pdEvent.HttpStatusCode.HasValue)
+                {
+                    if (!resourceData.StatusCodes.TryGetValue(pdEvent.HttpStatusCode.Value, out var count))
+                    {
+                        count = 0;
+                    }
+                    resourceData.StatusCodes[pdEvent.HttpStatusCode.Value] = count + 1;
+                }
+            }
+        }
+
+        private static void ApplyTiming(ResourceTimingData timingData, TimeSpan duration)
+        {
+            timingData.Requests++;
+            timingData.TotalDuration += duration;
+
+            if (timingData.MinDuration > duration)
+            {
+                timingData.MinDuration = duration;
+            }
+
+            if (timingData.MaxDuration < duration)
+            {
+                timingData.MaxDuration = duration;
+            }
+        }
+
+        public void Dispose()
+        {
+            ProtocolDiagnostics.Event -= ProtocolDiagnostics_Event;
+
+            var parentId = _parentId.ToString();
+            foreach (var kvp in _data)
+            {
+                Data data = kvp.Value;
+                string source = kvp.Key;
+                if (!_sources.TryGetValue(kvp.Key, out PackageSource packageSource))
+                {
+                    packageSource = null;
+                }
+
+                var telemetry = ToTelemetry(data, source, packageSource, parentId);
```

Comment:
Maybe put this in a flush/send method? 
I'd prefer the telemetry event to be fired explicitly rather than relying on dispose.